### PR TITLE
Add support for AES encryption and key wrap algorithms

### LINKED ISSUE #14887: Add oct-HSM support
For Key Vault 7.2-preview, make sure "oct" is already supported, and add "oct-HSM" support to models (names taken from swagger):

- [x] `KeyProperties.kty`
- [x] `KeyCreateParameters.kty`
- [x] `JsonWebKey.kty`

See the [service version 7.1 -> 7.2-preview diff](https://portal.azure-devex-tools.com/tools/diff/eyJzb3VyY2VVcmwiOiJodHRwczovL2dpdGh1Yi5jb20vQXp1cmUvYXp1cmUtcmVzdC1hcGktc3BlY3MvYmxvYi9tYXN0ZXIvc3BlY2lmaWNhdGlvbi9rZXl2YXVsdC9kYXRhLXBsYW5lL3JlYWRtZS5tZCIsInRhcmdldFVybCI6Imh0dHBzOi8vZ2l0aHViLmNvbS9BenVyZS9henVyZS1yZXN0LWFwaS1zcGVjcy9ibG9iL21hc3Rlci9zcGVjaWZpY2F0aW9uL2tleXZhdWx0L2RhdGEtcGxhbmUvcmVhZG1lLm1kIiwic291cmNlVGFnIjoicGFja2FnZS03LjEiLCJ0YXJnZXRUYWciOiJwYWNrYWdlLTcuMi1wcmV2aWV3In0=) 🔒 for details.

### LINKED ISSUE #14888: Add algorithms to EncryptionAlgorithm for Key Vault
For Key Vault 7.2-preview, to `EncryptionAlgorithm` (in swagger, the `KeyOperationsParameters` model), add the following algorithms:

- A128GCM
- A192GCM
- A256GCM
- A128KW
- A192KW
- A256KW
- A128CBC
- A192CBC
- A256CBC
- A128CBCPAD
- A192CBCPAD
- A256CBCPAD

As many of these as possible should be supported locally without writing our own cryptography.

Also add the following properties:

- `Iv` (model property: `iv`)
- `AdditionalAuthenticatedData` (model property: `aad`)
- `Tag` (model property: `tag`)

These will likely need to be in a new options type/parameter passed to `Encrypt` and `Decrypt`.

See the [service version 7.1 -> 7.2-preview diff](https://portal.azure-devex-tools.com/tools/diff/eyJzb3VyY2VVcmwiOiJodHRwczovL2dpdGh1Yi5jb20vQXp1cmUvYXp1cmUtcmVzdC1hcGktc3BlY3MvYmxvYi9tYXN0ZXIvc3BlY2lmaWNhdGlvbi9rZXl2YXVsdC9kYXRhLXBsYW5lL3JlYWRtZS5tZCIsInRhcmdldFVybCI6Imh0dHBzOi8vZ2l0aHViLmNvbS9BenVyZS9henVyZS1yZXN0LWFwaS1zcGVjcy9ibG9iL21hc3Rlci9zcGVjaWZpY2F0aW9uL2tleXZhdWx0L2RhdGEtcGxhbmUvcmVhZG1lLm1kIiwic291cmNlVGFnIjoicGFja2FnZS03LjEiLCJ0YXJnZXRUYWciOiJwYWNrYWdlLTcuMi1wcmV2aWV3In0=) 🔒 for details.

### PR DESCRIPTION
**Pull request by heaths**
Resolves #14887 and resolves #14888

### CODE REVIEW

File: sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/EncryptOptions.cs
Comment (heaths) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/EncryptOptions.cs:21]: @KrzysztofCwalina @tg-msft @schaabs Scott and I talked about this and originally we thought maybe an algorithm-specific class like `AesEncryptionOptions` (and make this class abstract) might help delineat options more, but as I was working on that it just wasn't very discoverable. By defining a constructor like this - including the optional param - I think we can better inform the user what to pass. If we should add other algorithm-specific variants in the future, we can add constructors.

I think this is better than, say, having an overload for each algorithm difference. In time, that will lead to many different overloads, though admittedly they should be fairly obvious which to use. Ultimately, too, internally we'll converge on a single request object but I know that's of little importance (i.e. just pointing out they all funnel to a single REST API anyway).

Thougths?

**Comment by heaths** [comment]
Note this is only for feedback. Once we have some initial consensus I'll add unit tests, work on adding recorded tests (we have an an environment I should be able to access), and leverage what algorithms we may from track 1 if they have been reviewed.

File: sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/CryptographyClient.cs
Comment (christothes) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/CryptographyClient.cs:142]: nit: We're inconsistent between this and "Encrypts plaintext". I actually like this one a bit better.

File: sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/CryptographyClient.cs
Comment (christothes) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/CryptographyClient.cs:291]: nit: Inconsistent with "Decrypts ciphertext". I think this one is more descriptive.

File: sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/CryptographyClient.cs
Comment (christothes) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/CryptographyClient.cs:166]: nit: The first sentence seems redundant. Suggestion applies to all the others as well.

```suggestion
        /// An <see cref="EncryptResult"/> containing the encrypted data
        /// along with all other information needed to decrypt it. This information should be stored with the encrypted data.
```

File: sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/LocalCryptographyClient.cs
Comment (christothes) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/LocalCryptographyClient.cs:114]: nite: inconsistent with "Encrypts the specified plain text."

File: sdk/keyvault/Azure.Security.KeyVault.Keys/src/KeyType.cs
Comment (christothes) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/src/KeyType.cs:58]: nit: If HSM here refers to Hardware Security Module, would "backed by an HSM" or "backed by a Hardware Security Module (HSM)" be better?

**Review by christothes — APPROVED** [review]
LGTM - Added some nits and minor suggestions.

File: sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/CryptographyClient.cs
Comment (heaths) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/CryptographyClient.cs:142]: That was intentional. You're not directly specifying the plaintext in the overload.

File: sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/CryptographyClient.cs
Comment (heaths) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/CryptographyClient.cs:291]: See above.

File: sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/CryptographyClient.cs
Comment (heaths) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/CryptographyClient.cs:166]: Thanks. I put that in some changes I'm working on already.

File: sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/LocalCryptographyClient.cs
Comment (heaths) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/LocalCryptographyClient.cs:114]: See above.

File: sdk/keyvault/Azure.Security.KeyVault.Keys/api/Azure.Security.KeyVault.Keys.netstandard2.0.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/api/Azure.Security.KeyVault.Keys.netstandard2.0.cs:364]: Since `iv` is optional should we make the property settable vs. setting via the ctor?

Comment (heaths) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/api/Azure.Security.KeyVault.Keys.netstandard2.0.cs:364]: /cc @KrzysztofCwalina because this is one I debated in my head a bit...

My thinking was that, while it's optional, it's only optional in some cases so putting it in the constructor helps hoist it to visibility. That said, I could do the same as I did for `DecryptOptions`, which was have overloads where - not only do I get a chance to document it*, but also can make sure that *that* ctor requires it (i.e. non-null).

\* That reminds me, I need to push an update with better docs to say which ctors to use with which algorithms. I'll wait for any of your additional feedback before I push, though.

Comment (heaths) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/api/Azure.Security.KeyVault.Keys.netstandard2.0.cs:364]: I was just about to change this and remembered why (given the current XML doc comment): yes, it can be null even when needed; however, in that case we will generate it. Making it an optional param gives me opportunity for a comment like so:
```c#
/// <param name="iv">
/// The initialization vector for encryption. If null, one will be generated for symmetric algorithms like an
/// nonce and returned in <see cref="EncryptResult.Iv"/>.
/// </param>
```
Yes, it could be done with two ctors but, the fact it's still optional in the other case would seem to obviate the need to have the overload. The overload that takes it could require it, but then people might think they are responsible for generating the 96-bit nonce for AES-GCM.

Comment (schaabs) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/api/Azure.Security.KeyVault.Keys.netstandard2.0.cs:364]: I'm assuming we generate in the client and not in the options class. If that's the case it seems we could just put this doc comment onto the property rather the the ctor param no?

Comment (heaths) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/api/Azure.Security.KeyVault.Keys.netstandard2.0.cs:364]: But it can also be passed in, and should when not used as an nonce.

That said, @KrzysztofCwalina agreed offline to factory methods like so, that will make these cases more clear:
```c#
EncryptResult Encrypt(EncryptOptions.A128CbcOptions(byte[] plaintext, byte[] iv), CancellationToken);
```

Comment (schaabs) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/api/Azure.Security.KeyVault.Keys.netstandard2.0.cs:364]: I like this approach. I'll approve this PR if we need to merge as is for a preview, but I think the approach you have with factory methods above is the best.

**Review by schaabs — APPROVED** [review]

**Pull request merged**
